virtcontainers/sandbox.go

```diff
@@ -633,6 +633,80 @@ func (s *Sandbox) IOStream(containerID, processID string) (io.WriteCloser, io.Re
 	return c.ioStream(processID)
 }
 
+// PreAddDevices can let sandbox attach devices to hypervisor before container created
+// it will be mainly used by frakti to boost startup time by doing this in parallel
+func (s *Sandbox) PreAddDevices(contConfig ContainerConfig) ([]api.Device, error) {
+	var (
+		c       *Container
+		devices []api.Device
+		err     error
+	)
+
+	defer func() {
+		if err != nil {
+			for _, d := range devices {
+				if rollbackErr := d.Detach(s); rollbackErr != nil {
+					s.Logger().WithError(rollbackErr).
+						WithFields(logrus.Fields{
+							"sandboxid": s.id,
+						}).Error("failed to detach devices for rolling back")
+				}
+			}
+			s.storage.deleteContainerResources(s.id, c.id, nil)
+		}
+	}()
+
+	c, err = newContainer(s, contConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = c.createContainersDirs(); err != nil {
+		return nil, err
+	}
+
+	// TODO: hotplug rootfs
```

Comment:
is there an issue for this?

Reply:
A lots of more work need to be done for this, so I added a TODO tag here.